docs/templates.md

```diff
@@ -601,3 +611,336 @@ access to the comments template tags and filters. The child template is
+or key to the translations. The interface to the resource bundle
+that Reinhardt expects is very simple:
+
+    var ResourceBundle = function(locale) { ... }
+    ResourceBundle.prototype.translate = function(source, key) { ... }
+
+That is, the constructor should load the translations (however it does that)
+and the translate method should return the translation for individual strings.
+Most strings will only have a source string, and the key will be undefined,
+because the source string already uniquely identifies the string.
+However, there is a way to
+explicitly specify a unique key for a string to distinguish it from other
+strings where the source is exactly the same, but which is used differently
+in in different contexts. (See below for more information on that.)
+
+You may use whatever translation package you want to implement this
+interface. Popular packages include ilib, i18n.js, or translate.js just to
+name a few. You do not have to re-invent the wheel for this, just pick
+a wheel that you think is nice and shiny and go with it.
+
+To set up a Reinhardt environment for translation, you should do something
+like this:
+
+    var env = new Environment({
+        i18n: {
+            resBundleFactory: function(locale) {
+                if (!this.cache) {
+                    this.cache = {};
+                }
+                if (!this.cache[locale]) {
+                    this.cache[locale] = new ResourceBundle(locale);
+                }
+                // resources are immutable, so we don't have to worry about sharing them
+                // amongst sessions and threads
+                return this.cache[locale];
+            },
+            defaultLocale: "en-US"
+        }
+    });
+
+This sets up the resource bundle factory to hand out new resource bundles for a locale
+when necessary. The example above also includes simple caching so that we don't have
+to reload the bundle every time. The environment needs a factory because each request
+that comes in may have a different locale, depending on the user making the request.
+
+That is almost everything you need to set up.
+
+The only thing left to discuss is how you render the response. Each request may come from
+a different user and therefore may have a different locale. The locale should be passed
+in to the rendering routines using the render context. Reinhardt can then use this
+locale to ask the resource bundle factory for a resource bundle for that locale, and
+then ask that resource bundle for translations to the strings it finds in the ``trans``
+and ``blocktrans`` tags. If no locale is given in the context, Reinhardt will use the
+default locale (if it is set up) or it will default to the text that is already in
+the templates.
+
+Here is an example of a route using stick that passes the locale in to Reinhardt
+in the render context:
+
+    app.get('/:page', function(request, page) {
+        return env.renderResponse(page, {
```

Comment:
> The problem is that the render() methods only take a context argument, so there is no way to pass it per-request information except via the context.

I would do just that: pass the locale as context.locale. The Environment doesn't have to be involved in i18n (execpt for holding i18n configuration): it just passes the context down to the Tags as it did before. And the Tags can lookup the resource bundle themselves. I admit it's not obvious that Tags can access the environment with `this.env`. But because of this, the `if` from Environment.renderResponse ( https://github.com/ehoogerbeets/reinhardt/blob/transtag/lib/environment.js#L50 ) can move into the render methods of BlockTransTag and TransTag.

If we do it like that, everything about i18n is contained in the i18n Tags. And passing locale via the context works in all cases: Env.render, Template.render and the convinience method Env.renderResponse. It makes `context.locale` a bit special but i can live with that.

Unrelated to the above:

>I'm not sure that it is thread-safe though. Is each request a separate thread? Is the context specific to a single thread so we don't have to worry about race conditions?

Yes, each request is in it's own thread and the context is not shared by threads. The docs warn about storing per-request information on Tag instances, if you are worried about that. Tags are used concurrently by multiple threads. i18n tags don't do that so they are fine.